src/ifcgeom/mapping/mapping.i

```diff
@@ -113,6 +113,9 @@ BIND(IfcEdge);
 BIND(IfcEdgeLoop);
 BIND(IfcPolyline);
 BIND(IfcPolyLoop);
+#ifdef SCHEMA_HAS_IfcGradientCurve
+BIND(IfcGradientCurve);
```

Comment:
@aothms This is a little problematic. IfcGradientCurve is a IfcCompositeCurve so the item gets processed twice and the result for IfcCompositeCurve is returned. When processed as IfCompositeCurve the IfcCurveSegment are the vertical segments.

I have some code in my driver program that maps a IfcGradientCurve but gets back the implicit_item for the vertical components.

```
if (item->as<Ifc4x3_add1::IfcGradientCurve>())
{
	auto mapped_item = ifcopenshell::geometry::taxonomy::cast<ifcopenshell::geometry::taxonomy::implicit_item>(mapping->map(item->as<Ifc4x3_add1::IfcGradientCurve>()));
	auto loop = ifcopenshell::geometry::taxonomy::cast<ifcopenshell::geometry::taxonomy::loop>(mapped_item->evaluate());
...
```
Reversing the order of BIND(IfcGradientCurve) and BIND(IfcCompositeCurve) fixes the problem, but I don't think it is the best solution.

Reply:
Oh sorry. Yes, that could be addressed like this: https://github.com/IfcOpenShell/IfcOpenShell/pull/3827/files#r1347391406

It's something I introduced in the implementation of caching. I could no longer use the return statement anymore (because writing to cache is at the bottom of the function). So now it "falls through" to the other. That's why my suggestion of the addition of `!item &&` helps circumvent that (well at least in the cases the mapping is successful in assigning something to `item`).

Reply:
@aothms This is still a little problematic. If `BIND(IfcCompositeCurve)` is listed before `BIND(IfcGradientCurve)` the `IfcGradientCurve `is treated as an `IfcCompositeCurve `and the `IfcGradientCurve `code never gets called.

Is there a function similar to  `inst->as<IfcSchema::T>()` that tests if the most derived type of an instance is equal to a specific type? 

For `IfcGradientCurve `the test for `IfcCompositeCurve `would be false since the most derived type is `IfcGradientCurve`. 

Listing the BIND statements for the more derived types before the parent types solves the problem, but this seems like a brittle and problematic solution.

Reply:
> Listing the BIND statements for the more derived types before the parent types solves the problem, but this seems like a brittle and problematic solution.

I agree, but in quite a few cases we do depend on matching also inherited types:

https://github.com/IfcOpenShell/IfcOpenShell/blob/661c8bc50f0a7226531a954e896bbe2cd971690c/src/ifcgeom/mapping/mapping.i#L33-L37

https://github.com/IfcOpenShell/IfcOpenShell/blob/661c8bc50f0a7226531a954e896bbe2cd971690c/src/ifcgeom/mapping/mapping.i#L103-L105

On the other hand in many other cases it's also problematic because in the inheritance hierarchy of the newly added IFC geometry types, they don't seem to consider liskov-substitution in determining whether entities are supposed to be subtypes, but instead rather pragmatically try to reuse some attribute definitions.

So maybe we need to enhance the BIND() macro with an option to selectively specify whether inheritance is considered to have the best of both worlds?

We could check equality of `&inst->declaration() == &Ifc2x3::IfcWall::Class()`. This disregards inheritance (and is also more efficient because it doesn't rely on RTTI and virtual inheritance).